src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs

```diff
@@ -776,7 +776,16 @@ internal ClassDataContractCriticalHelper(Type type) : base(type)
                 this.IsValueType = type.IsValueType;
                 if (baseType != null && baseType != Globals.TypeOfObject && baseType != Globals.TypeOfValueType && baseType != Globals.TypeOfUri)
                 {
+#if uapaot
+                    // The DataContact pre-generated by SG is imcomplete, in other words, some of its fields are not
+                    // populated by SG. Reflection-based serialization, however, requires a complete DataContract.
+                    // At this point, we're in the middle of creating a new DataContract at runtime, which indicates
+                    // reflection-based serialization is being used. Therefore, we cannot use the pre-generated DataContracts
+                    // of base types, even if they exist. 
+                    DataContract baseContract = DataContract.GetDataContractCreatedAtRuntime(baseType);
```

Comment:
Could we add a comment describing the "#if uapaot". Something along the lines of: "We can't use SG generated data contract as a base type for a data contract generated by this code. So we need to make sure that we create our own base type contract if we would potentially get here for a type with base type having SG generated data contract. Currently we don't have a good way of detecting usage of SG (either globally or per data contract). Since SG is currently only used by .NET Native, as a point-in-time solution we're using the "#if uapaot" to differentiate. Note that this code can be reached either from reflection based or ILGen based serializers. For uapaot, only reflection based serializers are possible (ILGen doesn't work in uapaot right now), so the ifdef actually means that we would only get here from reflection based serializers with possible usage of SG.